sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuGenerateExec.scala

```diff
@@ -264,38 +264,151 @@ abstract class GpuExplodeBase extends GpuUnevaluableUnaryExpression with GpuGene
       }
   }
 
+  private def getPerRowRepetition(explodingColumn: ColumnVector, outer: Boolean): ColumnVector = {
+    withResource(explodingColumn.countElements()) { arrayElements =>
+      if (outer) {
+        // for outer, empty arrays and null arrays will produce a row
+        withResource(GpuScalar.from(1, IntegerType)) { one =>
+          val noNulls = withResource(arrayElements.isNotNull) { isLhsNotNull =>
+            isLhsNotNull.ifElse(arrayElements, one)
+          }
+          withResource(noNulls) { _ =>
+            withResource(noNulls.greaterOrEqualTo(one)) { isGeOne =>
+              isGeOne.ifElse(noNulls, one)
+            }
+          }
+        }
+      } else {
+        // ensure no nulls in this output
+        withResource(GpuScalar.from(0, IntegerType)) { zero =>
+          GpuNvl(arrayElements, zero)
+        }
+      }
+    }
+  }
+
+  private def getRowByteCount(column: Seq[ColumnVector]): ColumnVector = {
+    val repeatTable = new Table(column:_*)
+    val bits = withResource(repeatTable) {
+      _.rowBitCount()
+    }
+    val bitsLong = withResource(bits) {
+      _.castTo(DType.INT64)
+    }
+    withResource(bitsLong) { _ =>
+      withResource(ai.rapids.cudf.Scalar.fromLong(8)) { toBytes =>
+        bitsLong.div(toBytes)
+      }
+    }
+  }
+
   override def inputSplitIndices(inputBatch: ColumnarBatch,
-      generatorOffset: Int,
-      outer: Boolean,
+      generatorOffset: Int, outer: Boolean,
       targetSizeBytes: Long): Array[Int] = {
 
-    val vectors = GpuColumnVector.extractBases(inputBatch)
     val inputRows = inputBatch.numRows()
     if (inputRows == 0) return Array()
 
+    val vectors = GpuColumnVector.extractBases(inputBatch)
+
+    val explodingColumn = vectors(generatorOffset)
+
     // Get the output size in bytes of the column that we are going to explode
     // along with an estimate of how many output rows produced by the explode
-    val (explodeColOutputSize, estimatedOutputRows) = withResource(
-      vectors(generatorOffset).getChildColumnView(0)) { listValues =>
-      val totalSize = listValues.getDeviceMemorySize
-      val totalCount = listValues.getRowCount
-      (totalSize.toDouble, totalCount.toDouble)
+    val (explodeColOutputSize, estimatedOutputRows) =
+      withResource(explodingColumn.getChildColumnView(0)) { listValues =>
+        val totalSize = explodingColumn.getDeviceMemorySize
+        // get the number of elements in the array child
+        var totalCount = listValues.getRowCount + listValues.getNullCount
```

Comment:
"the number of elements in the array child" should be just `listValues.getRowCount`, which already covered null count and non-null count.

Reply:
thanks for catching that, you are right. I think I need to invert the logic a bit then.

Reply:
Fixed in: https://github.com/NVIDIA/spark-rapids/pull/6631/commits/c8288aa3e37380aab93a6dd42ffea7d195da4c86.